src/simulations/collect.jl

```diff
@@ -0,0 +1,285 @@
+## Basic data collection functions
+
+get_data(a, s::Symbol) = getproperty(a, s)
+get_data(a, f::Function) = f(a)
+
+"""
+    collect_agent_data(model::ABM, properties::AbstractArray, step::Int)
+
+Collect agent properties into a dataframe. `properties` can have symbols (agent fields) or functions that take an agent as input.
+"""
+function collect_agent_data(model::ABM, properties::AbstractArray, step::Int)
+  dd = DataFrame()
+  dd[!, :id] = collect(keys(model.agents))
+  for fn in properties
+    dd[!, fn] = get_data.(values(model.agents), fn)
+  end
+  dd[!, :step] = fill(step, size(dd, 1))
+  return dd
+end
+
+"""
+    collect_model_data(model::ABM, properties::AbstractArray, step::Int)
+
+Collect model properties from functions or symbols provided in `properties`.
+"""
+function collect_model_data(model::ABM, properties::AbstractArray, step::Int)
+  dd = DataFrame()
+  for fn in properties
+    r =  get_data(model, fn)
+    dd[!, Symbol(fn)] = [r]
+  end
+  dd[!, :step] = fill(step, size(dd, 1))
+  return dd
+end
+
+"""
+    aggregate_data(df::AbstractDataFrame, aggregation_dict::Dict)
+  
+Aggregate `df` columns  with some function(s) specified in `aggregation_dict`.
+Each key in `aggregation_dict` is a column name (Symbol), and each value is
+an array of function to aggregate that column.
+
+Aggregation occurs per step.
+"""
+function aggregate_data(df::AbstractDataFrame, aggregation_dict::Dict)
+  all_keys = collect(keys(aggregation_dict))
+  dfnames = names(df)
+  available_keys = [k for k in all_keys if in(k, dfnames)]
+  length(available_keys) == 0 && return
+  v1 = aggregation_dict[available_keys[1]]
+  final_df = by(df, :step,  available_keys[1] => v1[1])
+  for v2 in v1[2:end]
+    dd = by(df, :step,  available_keys[1] => v2)
+    final_df = join(final_df, dd, on=:step)
+  end
+  for k in available_keys[2:end]
+    v = aggregation_dict[k]
+    for v2 in v
+      dd = by(df, :step,  k => v2)
+      final_df = join(final_df, dd, on=:step)
+    end
+  end
+
+  # rename columns
+  colnames = Array{Symbol}(undef, size(final_df, 2)) 
+  colnames[1] = :step
+  counter = 2
+  for k in available_keys
+    v = aggregation_dict[k]
+    for vv in v
+      colnames[counter] = Symbol(join([vv,"(", string(k), ")"], ""))
+      counter += 1
+    end
+  end
+  rename!(final_df, colnames)
+
+  return final_df
+end
+
+## step functions
+###################
+
+"used in _step!"
+function _step_agent_collect!(df::DataFrame, model::ABM, agent_properties,  aggregation_dict, step::Int)
+  if isnothing(aggregation_dict)
+    dft = collect_agent_data(model, agent_properties, step)
+    df = vcat(df, dft)
+  else
+    dfall = collect_agent_data(model, agent_properties, step)
```

Comment:
Line `dft = ...` and `dfall = ...` is identical in both clauses. Please put it outside the `if` clause to reduce code duplication.